### PR TITLE
Stop an infinite loop when using an exported JSDoc function as a type

### LINKED ISSUE #36830: JSDOC @type tag crashes with simple export
**TypeScript Version:**  3.9.0-dev.20200215 (also tested 3.7.5)

(the last one that doesn't crash is `v3.7.1-rc`. `v3.7.2` crashes)

**Search Terms:**

`Maximum call stack size exceeded`
`getTypeOfFuncClassEnumModule`

**Code**

`/lib/index.js`:
```js
/** @type {import('../types/index').xxx} */
function xxx() {
    return null;
}
```

`/types/index.d.ts`:
```ts
export function xxx(): string;
```

`/tsconfig.json`:
```js
{
    "compilerOptions": {
        "target": "esnext",
        "module": "commonjs",
        "esModuleInterop": true,
        "lib": [
            "dom",
            "esnext"
        ],
        "allowJs": true,
        "checkJs": true,
        "noEmit": true,
        "resolveJsonModule": true,
        "strict": true,
        "moduleResolution": "node",
        "baseUrl": ".",
    }
}
```

**Expected behavior:**

No crash.
Typescript error about function returning wrong error type.

**Actual behavior:**

Crash.

```
/me/ts-crash-test/node_modules/typescript/lib/tsc.js:81193
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at getSymbolLinks (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:28829:32)
    at getTypeOfFuncClassEnumModule (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:33936:25)
    at getTypeOfSymbol (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:34045:24)
    at getTypeFromJSDocValueReference (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36634:33)
    at getTypeReferenceType (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36620:33)
    at getTypeFromJSDocValueReference (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36648:36)
    at getTypeReferenceType (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36620:33)
    at getTypeFromJSDocValueReference (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36648:36)
    at getTypeReferenceType (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36620:33)
    at getTypeFromJSDocValueReference (/me/ts-crash-test/node_modules/typescript/lib/tsc.js:36648:36)
```

**Playground Link:** 

https://github.com/rchl/ts-crash-test

**Related Issues:**

### PR DESCRIPTION
**Pull request by orta**
Fixes #36830

In JS we want to support treating values as types in a bunch of positions, the code in https://github.com/microsoft/TypeScript/pull/35057 added the ability for this to work in classes but it was over-reaching and also getting called when you have a function. This caused an infinite loop as the type lookup was the same object and not something which could differ in the way an instance vs the 'class' (e.g. the static stuff) could differ.

### CODE REVIEW

File: tests/baselines/reference/jsdocImportTypeReferenceToClassAlias.types [hunk 1]
```diff
@@ -19,11 +19,11 @@ module.exports.C = C
...
 function demo(c) {
 >demo : (c: X) => void
->c : C
+>c : typeof C
 
     c.s
->c.s : () => void
->c : C
->s : () => void
+>c.s : any
```
Comment (sandersn) [inline tests/baselines/reference/jsdocImportTypeReferenceToClassAlias.types:25]: this seems wrong, which means that this fix may also need to be restricted to types that come through `require`

**Comment by sandersn** [comment]
@weswigham 
The intent for this code is to de-alias fake JS aliases to types.

1. The old code was unconditional and non-obvious -- it always assumed the symbol was an alias, dereferenced it, then if dereferencing changed the symbol, returned it. This overapplied in non-obvious places.
2. The new code is syntactic and explicit, so it underapplies -- although it only needs to handle two cases right now. I expect a tail of bugs from it, but I think the result will be safer. In this case, I didn't understand that only classes need the additional dereference step since they're incorrectly imported via `require` as value-only. 

tl;dr: There are two problems: (1) an unknown number of cases that were previously being handled (2) the existing cases are tricky for me to understand.

3. I think you are right. Here's a possible example:

```ts
// @filename: lib/index.js
/** @type {import('../folder/index').zzz} */
export function xxx() {
  return 12;
}

// @filename: folder/index.d.ts
export function zzz(): string;
export type zzz = (n: number) => number;
```

The intent is for `xxx` to have type `(n: number) => number`, but today it would not.

4. Overall, we could get rid of this bug farm by making `require` produce real aliases. Well, at least we could relocate it to a good neighbourhood and  clean it up. That needs a couple up-front tasks:

    a. We should figure out how worthwhile high-quality `require` support is. Probably worth it' since the 2018 survey we've seen that people write ES imports twice as much as commonjs, but I don't have an idea of how much commonjs we support in absolute terms.
    b. I need to relearn the details of binding of aliases, real and fake, because I'm not sure what work is required to make this happen.

**Comment by orta** [comment]
Deprecated by https://github.com/microsoft/TypeScript/pull/39770
